### PR TITLE
chore: replace Ring with RustCrypto in spki.rs

### PR DESCRIPTION
**Pull request by rjzak**

### CODE REVIEW

File: src/crypto/spki.rs [hunk 1]
```diff
@@ -46,44 +44,50 @@ pub trait SubjectPublicKeyInfoExt {
...
-                        let p = p.decode_into::<AlgorithmIdentifier<'_>>()?;
-                        match (p.oids()?, salt, tfld) {
-                            ((SHA256, None), 32, 1) => Ok(SHA256),
-                            ((SHA384, None), 48, 1) => Ok(SHA384),
-                            ((SHA512, None), 64, 1) => Ok(SHA512),
-                            _ => Err(anyhow!("unsupported")),
-                        }
-                    }
-                    _ => Err(anyhow!("unsupported")),
-                }?;
```
Comment (npmccallum) [inline src/crypto/spki.rs:73]: It is important not to remove this check.

File: src/crypto/spki.rs [hunk 2]
```diff
@@ -46,44 +44,50 @@ pub trait SubjectPublicKeyInfoExt {
...
-        let alg: &'static dyn VerAlg = match (self.algorithm.oids()?, (algo.oid, algo.parameters)) {
-            ((ECPK, Some(P256)), ES256) => &ECDSA_P256_SHA256_ASN1,
-            ((ECPK, Some(P384)), ES384) => &ECDSA_P384_SHA384_ASN1,
+        match (self.algorithm.oids()?, (algo.oid, algo.parameters)) {
+            ((ECPK, Some(P256)), ES256) => {
+                use p256::ecdsa::signature::Verifier;
+                let vkey = p256::ecdsa::VerifyingKey::from_sec1_bytes(self.subject_public_key)?;
+                let sig = p256::ecdsa::Signature::from_der(sign)?;
+                return Ok(vkey.verify(&body, &sig)?);
+            }
```
Comment (npmccallum) [inline src/crypto/spki.rs:53]:
```suggestion
            }

```

File: src/crypto/spki.rs [hunk 3]
```diff
@@ -46,44 +44,50 @@ pub trait SubjectPublicKeyInfoExt {
...
+            ((ECPK, Some(P384)), ES384) => {
+                use p384::ecdsa::signature::Verifier;
+                let vkey = p384::ecdsa::VerifyingKey::from_sec1_bytes(self.subject_public_key)?;
+                let sig = p384::ecdsa::Signature::from_der(sign)?;
+                return Ok(vkey.verify(&body, &sig).unwrap());
+            }
```
Comment (npmccallum) [inline src/crypto/spki.rs:59]:
```suggestion
            }

```

**Review by npmccallum — CHANGES_REQUESTED** [review]

File: src/crypto/spki.rs (hunk 1 above)
Comment (rjzak) [reply src/crypto/spki.rs:73]: Which check is that? The outer match (EC vs RSA) has the default case, the inner match for SHA hashes for RSA has the default case, and the RSA mask.parameters has the `ok_or()` check.

File: src/crypto/spki.rs (hunk 2 above)
Comment (rjzak) [reply src/crypto/spki.rs:53]: Done

File: src/crypto/spki.rs (hunk 3 above)
Comment (rjzak) [reply src/crypto/spki.rs:60]: Done

File: src/crypto/spki.rs (hunk 1 above)
Comment (npmccallum) [reply src/crypto/spki.rs:73]: We are also checking the size of the `salt` and `trailer`.

Basically, ID_MGF_1 allows signers to specify crazy parameters. We don't. We require only sane combinations.

For example, we support `SHA256` with no parameters, a `salt_length` of `32` and a `trailer_field` of `1`. We reject all other variants of `SHA256`.

**Review by npmccallum — CHANGES_REQUESTED** [review]

File: src/crypto/spki.rs (hunk 1 above)
Comment (npmccallum) [reply src/crypto/spki.rs:73]: You should basically keep this match statement as it exists today. It isn't `ring` specific but rather `X.509` specific.

Comment (rjzak) [reply src/crypto/spki.rs:73]: It's back

**Review by npmccallum — APPROVED** [review]

**Pull request merged**